### PR TITLE
Change resource.New() to use functional options; add builtin attributes for (host.*, telemetry.sdk.*)

### LINKED ISSUE #1076: SDK should populate telemetry.sdk.* resource attributes
As per the [spec](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/resource/semantic_conventions/README.md#telemetry-sdk), the SDK must set the `opentelemetry.sdk.name` to `opentelemetry`. It would be nice if the other `telemetry.sdk.*` values were also set.

### PR DESCRIPTION
**Pull request by jmacd**
This adds `resource.NewConfig(ctx, ...)` with functional options:
- `WithAttributes(...)` for directly adding key:values
- `WithDetectors(...)` for directly adding resource detectors
- `WithTelemetrySDK(...)` to override the default `telemetry.sdk.*` resources
- `WithHost(...)` to override the default `host.*` resources
- `WithFromEnv(...)` to override the default [environment configuration](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/resource/sdk.md#detecting-resource-information-from-the-environment)
- `WithoutBuiltin()` to disable all builtin resources

The intention is to make it easy to configure a `*Resource` with the standard resources. This adds standard resource detectors for the telemetry SDK and host name.

Resolves #1076.

### CODE REVIEW

**Comment by jmacd** [comment]
Resolves #1076

**Review by Aneurysm9 — APPROVED** [review]

File: sdk/resource/builtin.go
Comment (jmacd) [inline sdk/resource/builtin.go:46]: nit: this should return a _partial_ error to not block startup.

Comment (XSAM) [reply sdk/resource/builtin.go:46]: Maybe we should use `ErrorHandler`?

File: sdk/resource/config_test.go
Comment (XSAM) [inline sdk/resource/config_test.go:31]: Nit: env variable might potentially affect other tests. Should we reset the env after this case?

BTW, we have an env helper called `EnvStore` in `internal/testing`, and it can record/set/restore env in a convenient way.

File: sdk/resource/config.go
Comment (MrAlias) [inline sdk/resource/config.go:117]: What are your thoughts on calling this `WithoutBuiltin` similar to the [bool option design](https://github.com/open-telemetry/opentelemetry-go/blob/master/CONTRIBUTING.md#bool-options)?

File: sdk/resource/config.go
Comment (MrAlias) [inline sdk/resource/config.go:131]: The lastest design shift has moved to calling this `NewConfig`: https://github.com/open-telemetry/opentelemetry-go/blob/master/CONTRIBUTING.md#config

File: sdk/metric/sdk.go
Comment (MrAlias) [inline sdk/metric/sdk.go:308]: In the tracing SDK the resource is associated at the Provider level. Is the idea that this will come from the provider, but allows this to be configured for testing?

**Review by MrAlias — APPROVED** [review]

**Comment by jmacd** [comment]
FYI I will address the feedback in this PR today.

File: sdk/resource/builtin.go
Comment (jmacd) [reply sdk/resource/builtin.go]: I added tests for error handling, but had misunderstood _partial_ error support. I would say that a new issue should be filed about `ErrorHandler` support in `resource.Detect()`. That code does not call the global error handler, but I think it should.

While we're at it, we might consider improving `resource.Detect()` not to block startup indefinitely, with a context timeout, and to run detectors in parallel.

File: sdk/resource/config_test.go
Comment (jmacd) [reply sdk/resource/config_test.go]: Fixed. I also updated the other tests in this package to `Restore()` the environment.

File: sdk/resource/config.go
Comment (jmacd) [reply sdk/resource/config.go]: Done.

File: sdk/resource/config.go
Comment (jmacd) [reply sdk/resource/config.go]: Done.

File: sdk/resource/config.go
Comment (jmacd) [reply sdk/resource/config.go]: Fixed.

File: sdk/resource/builtin.go
Comment (MrAlias) [inline sdk/resource/builtin.go:29]: Can we add some docs to the exported types and functions?

```suggestion
	// TelemetrySDK is a Detector that provides information about the OpenTelemetry SDK used.
	// This Detector is included as a builtin. If this information is not wanted it needs to explicitly be
	// configured to not be included.
	TelemetrySDK struct{}
	// Host is a Detector that provides information about the host being run on. This Detector is
	// included as a builtin. If this information is not wanted it needs to explicitly be configured to
	// not be included.
	Host         struct{}
```

File: sdk/resource/builtin.go
Comment (MrAlias) [inline sdk/resource/builtin.go:43]:
```suggestion
// Detect returns a *Resource that describes the OpenTelemetry SDK used.
func (TelemetrySDK) Detect(context.Context) (*Resource, error) {
```

File: sdk/resource/builtin.go
Comment (MrAlias) [inline sdk/resource/builtin.go:51]:
```suggestion
// Detect returns a *Resource that describes the host being run on.
func (Host) Detect(ctx context.Context) (*Resource, error) {
```

File: sdk/resource/builtin.go
Comment (MrAlias) [inline sdk/resource/builtin.go:55]:
```suggestion
// StringDetector returns a Detector that will produce a *Resource containing the string as a value corresponding to k.
func StringDetector(k label.Key, f func() (string, error)) Detector {
```

File: sdk/resource/config.go
Comment (MrAlias) [inline sdk/resource/config.go:131]: I don't think I was paying too close attention in my last review when I suggested we rename this from `Configure` to `NewConifg`. This doesn't return a `config`, to that probably not the best name. I'm guessing a better name would reflect the `Resource` being returned and the use of `Detectors` to create it. Possibly `NewFromDetectors` to compliment `New`.

I don't want the naming to block this PR. I'll open an issue to track the renaming.

File: sdk/resource/env_test.go
Comment (MrAlias) [inline sdk/resource/env_test.go:30]: Thanks for cleaning this up :heart:

**Review by MrAlias — APPROVED** [review]
We can probably merge this as is. Adding the exported type/method docs would be nice, but could be added in a follow on.

File: sdk/resource/config.go
Comment (MrAlias) [reply sdk/resource/config.go:131]: https://github.com/open-telemetry/opentelemetry-go/issues/1263

**Comment by jmacd** [comment]
I will document this PR today. Sorry @MrAlias, and thanks for keeping me honest.

File: sdk/resource/config.go
Comment (jmacd) [reply sdk/resource/config.go:136]: OK I'm happy with `NewFromDetectors`, otherwise this PR is ready to merge.

**Review by XSAM — APPROVED** [review]

**Review by Aneurysm9 — APPROVED** [review]

**Comment by jmacd** [comment]
@open-telemetry/go-approvers how would you feel about removing the current `resource.New(...label.KeyValue) *Resource` and renaming this PR's `NewConfig()` as the replacement API, i.e., `resource.New(context.Context, ...Option) (*Resource, error)`?  I think this will encourage the use of the detector pattern.

**Comment by MrAlias** [comment]
> @open-telemetry/go-approvers how would you feel about removing the current `resource.New(...label.KeyValue) *Resource` and renaming this PR's `NewConfig()` as the replacement API, i.e., `resource.New(context.Context, ...Option) (*Resource, error)`? I think this will encourage the use of the detector pattern.

I am in favor of this :+1:

**Comment by XSAM** [comment]
> how would you feel about removing the current resource.New(...label.KeyValue) *Resource and renaming this PR's NewConfig() as the replacement API, i.e., resource.New(context.Context, ...Option) (*Resource, error)? I think this will encourage the use of the detector pattern.

I'm not sure about this. `New(kvs ...label.KeyValue) *Resource` seems quite simple to me. Also, does any chance that `resource.New` might return an error?

**Comment by jmacd** [comment]
> I'm not sure about this.

I had originally used `Configure()` as the name, trying to make it clear that `Configure()` had some logic (possible errors) whereas `New()` just builds a resource (no errors). I hope that most end users will call only a single method, is the thing added in this PR, and the current `New()` method will mostly be used by detector implementations. Whatever we do, we should choose the right name from the majority of users' perspective.

**Comment by Aneurysm9** [comment]
> @open-telemetry/go-approvers how would you feel about removing the current `resource.New(...label.KeyValue) *Resource` and renaming this PR's `NewConfig()` as the replacement API, i.e., `resource.New(context.Context, ...Option) (*Resource, error)`? I think this will encourage the use of the detector pattern.

I think that I like this approach.  The addition of `resource.WithAttributes()` seems like a pretty small overhead for a more flexible API that may also encourage the use of detectors.

**Comment by jmacd** [comment]
I'd like this to make progress. The current ideas for this signature:
- NewConfig
- Configure
- New (where existing New is renamed something else).

Another proposal is to change the signature of `Detect()` to support the options in this PR. Most users, then, would only call `Detect()`.

**Comment by Aneurysm9** [comment]
I think my order of preference would be:

* New (with existing New renamed or dropped)
* Detect
* NewConfig
* Configure

with the caveat that I for some reason actively dislike the last two.  They just don't feel right compared with the other options.

**Comment by jmacd** [comment]
I applied your suggestion @Aneurysm9, renamed the new method to `New()` and renamed the old method to `NewFromAttributes()`.

**Review by MrAlias — APPROVED** [review]

**Comment by MrAlias** [comment]
@XSAM we talked about this issue in the SIG meeting today and wanted to make sure we have consensus on this change going forward. We definitely value your opinion and want to understand if you are in strong opposition to these changes and if we can find an agreeable solution.

**Comment by XSAM** [comment]
@MrAlias I misunderstand what @jmacd is proposed. After reviewing the implementation, I believe this is the right decision.

**Review by XSAM — APPROVED** [review]

**Comment by jmacd** [comment]
The only last thought I have here is maybe the new method should be `NewWithAttributes()` instead of `NewFromAttributes()`, since it's an abbreviation for `New(..., WithAttributes(...))` that discards a nil error.

**Comment by MrAlias** [comment]
> The only last thought I have here is maybe the new method should be `NewWithAttributes()` instead of `NewFromAttributes()`, since it's an abbreviation for `New(..., WithAttributes(...), WithoutBuiltin())` that discards a nil error.

Yeah, that makes sense.

**Comment by jmacd** [comment]
I think this is ready! 😀

**Pull request merged**
